transports/quic/src/transport.rs

```diff
@@ -0,0 +1,780 @@
+            .field("pending_event", &self.pending_event)
+            .finish()
+    }
+}
+
+impl<P: Provider> Drop for Listener<P> {
+    fn drop(&mut self) {
+        self.endpoint_channel.send_on_drop(ToEndpoint::Decoupled);
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum SocketFamily {
+    Ipv4,
+    Ipv6,
+}
+
+impl SocketFamily {
+    fn is_same(a: &IpAddr, b: &IpAddr) -> bool {
+        matches!(
+            (a, b),
+            (IpAddr::V4(_), IpAddr::V4(_)) | (IpAddr::V6(_), IpAddr::V6(_))
+        )
+    }
+}
+
+impl From<IpAddr> for SocketFamily {
+    fn from(ip: IpAddr) -> Self {
+        match ip {
+            IpAddr::V4(_) => SocketFamily::Ipv4,
+            IpAddr::V6(_) => SocketFamily::Ipv6,
+        }
+    }
+}
+
+/// Turn an [`IpAddr`] reported by the interface watcher into a
+/// listen-address for the endpoint.
+///
+/// For this, the `ip` is combined with the port that the endpoint
+/// is actually bound.
+///
+/// Returns `None` if the `ip` is not the same socket family as the
+/// address that the endpoint is bound to.
+fn ip_to_listenaddr(endpoint_addr: &SocketAddr, ip: IpAddr) -> Option<Multiaddr> {
+    // True if either both addresses are Ipv4 or both Ipv6.
+    if !SocketFamily::is_same(&endpoint_addr.ip(), &ip) {
+        return None;
+    }
+    let socket_addr = SocketAddr::new(ip, endpoint_addr.port());
+    Some(socketaddr_to_multiaddr(&socket_addr))
+}
+
+/// Tries to turn a QUIC multiaddress into a UDP [`SocketAddr`]. Returns None if the format
+/// of the multiaddr is wrong.
+fn multiaddr_to_socketaddr(addr: &Multiaddr) -> Option<SocketAddr> {
+    let mut iter = addr.iter();
+    let proto1 = iter.next()?;
+    let proto2 = iter.next()?;
+    let proto3 = iter.next()?;
+
+    for proto in iter {
+        match proto {
+            Protocol::P2p(_) => {} // Ignore a `/p2p/...` prefix of possibly outer protocols, if present.
+            _ => return None,
+        }
+    }
+
+    match (proto1, proto2, proto3) {
+        (Protocol::Ip4(ip), Protocol::Udp(port), Protocol::QuicV1) => {
+            Some(SocketAddr::new(ip.into(), port))
+        }
+        (Protocol::Ip6(ip), Protocol::Udp(port), Protocol::QuicV1) => {
+            Some(SocketAddr::new(ip.into(), port))
+        }
+        _ => None,
+    }
+}
+
+/// Whether an [`Multiaddr`] is a valid for the QUIC transport.
+fn is_quic_addr(addr: &Multiaddr) -> bool {
+    use Protocol::*;
+    let mut iter = addr.iter();
+    let first = match iter.next() {
+        Some(p) => p,
+        None => return false,
+    };
+    let second = match iter.next() {
+        Some(p) => p,
+        None => return false,
+    };
+    let third = match iter.next() {
+        Some(p) => p,
+        None => return false,
+    };
+    let fourth = iter.next();
+    let fifth = iter.next();
+
+    matches!(first, Ip4(_) | Ip6(_) | Dns(_) | Dns4(_) | Dns6(_))
+        && matches!(second, Udp(_))
+        && matches!(third, QuicV1)
```

Comment:
Right now we only support the new `quic-v1` codepoint. Support for `quic`/ draft-29 will be added in a follow-up PR. See https://github.com/multiformats/multiaddr/issues/145#issuecomment-1309196984 for more info. cc @marten-seemann